File: brain/sdk/typescript/src/brain.ts
/**
 * Brain SDK - Main Brain Class
 *
 * Usage:
 *   import { Brain } from '@workspace/brain';
 *   const brain = Brain.load();
 *
 *   // Query beliefs
 *   if (brain.believes('distribution-beats-product')) { ... }
 *
 *   // Get network
 *   const experts = brain.network.domainMatches('sales');
 */

import {
  BrainContext,
  BrainState,
  Entity,
  Belief,
  Thread,
  Relationship,
  Prediction,
  Agenda,
  Network,
  Connection,
  ConfidenceLevel,
  RelationshipStrength,
} from './types';

import {
  loadEntities,
  loadRelationships,
  loadPredictions,
  loadAgenda,
  loadNetwork,
  loadState,
} from './loaders';

export class Brain {
  private _state: BrainState;
  private _entities: Entity[];
  private _relationships: Relationship[];
  private _predictions: Prediction[];
  private _agenda: Agenda;
  private _network: Network;

  private constructor(
    state: BrainState,
    entities: Entity[],
    relationships: Relationship[],
    predictions: Prediction[],
    agenda: Agenda,
    network: Network,
  ) {
    this._state = state;
    this._entities = entities;
    this._relationships = relationships;
    this._predictions = predictions;
    this._agenda = agenda;
    this._network = network;
  }

  /**
   * Load the brain from disk
   */
  static load(): Brain {
    const state = loadState();
    const entities = loadEntities();
    const relationships = loadRelationships();
    const predictions = loadPredictions();
    const agenda = loadAgenda();
    const network = loadNetwork();

    return new Brain(state, entities, relationships, predictions, agenda, network);
  }

  // === STATE ===

  get state(): BrainState {
    return this._state;
  }

  get version(): string {
    return this._state.version;
  }

  get capabilities(): Record<string, boolean> {
    return this._state.capabilities;
  }

  hasCapability(name: string): boolean {
    return this._state.capabilities[name] === true;
  }

  // === ENTITIES ===

  get entities(): Entity[] {
    return this._entities;
  }

  get beliefs(): Belief[] {
    return this._entities.filter(e => e.type === 'belief') as Belief[];
  }

  get threads(): Thread[] {
    return this._entities.filter(e => e.type === 'thread') as Thread[];
  }

  entity(id: string): Entity | undefined {
    return this._entities.find(e => e.id === id);
  }

  /**
   * Check if the brain holds a belief
   */
  believes(beliefId: string): boolean {
    const belief = this.entity(`belief.${beliefId}`) || this.entity(beliefId);
    return belief !== undefined;
  }

  /**
   * Get belief confidence level
   */
  beliefConfidence(beliefId: string): ConfidenceLevel | undefined {
    const belief = this.entity(`belief.${beliefId}`) || this.entity(beliefId);
    return belief?.confidence;
  }

  /**
   * Get entities by type
   */
  entitiesByType(type: Entity['type']): Entity[] {
    return this._entities.filter(e => e.type === type);
  }

  /**
   * Get entities by confidence level
   */
  entitiesByConfidence(confidence: ConfidenceLevel): Entity[] {
    return this._entities.filter(e => e.confidence === confidence);
  }

  // === RELATIONSHIPS ===

  get relationships(): Relationship[] {
    return this._relationships;
  }

  /**
   * Get relationships for an entity
   */
  relationshipsFor(entityId: string): Relationship[] {
    return this._relationships.filter(r => r.from === entityId || r.to === entityId);
  }

  /**
   * Get what supports an entity
   */
  supports(entityId: string): Relationship[] {
    return this._relationships.filter(r => r.to === entityId && r.type === 'supports');
  }

  /**
   * Get what contradicts an entity
   */
  contradicts(entityId: string): Relationship[] {
    return this._relationships.filter(r => r.to === entityId && r.type === 'contradicts');
  }

  // === PREDICTIONS ===

  get predictions(): Prediction[] {
    return this._predictions;
  }

  /**
   * Get pending predictions
   */
  pendingPredictions(): Prediction[] {
    return this._predictions.filter(p => p.status === 'pending');
  }

  /**
   * Get predictions due soon
   */
  predictionsDueBefore(date: Date): Prediction[] {
    const dateStr = date.toISOString().split('T')[0];
    return this._predictions.filter(p =>
      p.status === 'pending' && p.resolution_date <= dateStr
    );
  }

  // === AGENDA ===

  get agenda(): Agenda {
    return this._agenda;
  }

  /**
   * Get high-priority agenda items
   */
  urgentAgendaItems(): Agenda['immediate'] {
    return this._agenda.immediate.filter(item => item.priority === 'high');
  }

  // === NETWORK ===

  get network(): NetworkAccessor {
    return new NetworkAccessor(this._network);
  }

  // === CONVENIENCE ===

  /**
   * Get attention items that need immediate focus
   */
  get pendingAttention(): string[] {
    return this._state.pending_attention || [];
  }

  /**
   * Get recent changes to the brain
   */
  get recentChanges(): string[] {
    return this._state.recent_changes || [];
  }
}

/**
 * Network accessor for convenient querying
 */
class NetworkAccessor {
  private _network: Network;

  constructor(network: Network) {
    this._network = network;
  }

  get connections(): Connection[] {
    return this._network.connections || [];
  }

  get stats(): Network['stats'] {
    return this._network.stats;
  }

  /**
   * Find connections in a domain
   */
  domainMatches(domain: string): Connection[] {
    const domainLower = domain.toLowerCase();
    return this.connections.filter(c =>
      c.domains?.some(d => d.toLowerCase().includes(domainLower))
    );
  }

  /**
   * Find connections by relationship strength
   */
  byStrength(strength: RelationshipStrength): Connection[] {
    return this.connections.filter(c => c.relationship_strength === strength);
  }

  /**
   * Get high-trust connections
   */
  highTrust(): Connection[] {
    return this.connections.filter(c => c.trust_level === 'high');
  }

  /**
   * Get energizing connections
   */
  energizing(): Connection[] {
    return this.connections.filter(c => c.energy === 'energizing');
  }

  /**
   * Get draining connections
   */
  draining(): Connection[] {
    return this.connections.filter(c => c.energy === 'draining');
  }

  /**
   * Get a specific connection
   */
  get(id: string): Connection | undefined {
    return this.connections.find(c => c.id === id);
  }

  /**
   * Search connections by name
   */
  search(query: string): Connection[] {
    const queryLower = query.toLowerCase();
    return this.connections.filter(c =>
      c.name.toLowerCase().includes(queryLower) ||
      c.company?.toLowerCase().includes(queryLower)
    );
  }
}

File: brain/sdk/python/brain/brain.py
"""
Brain SDK - Main Brain Class

Usage:
    from brain import Brain

    brain = Brain.load()

    # Check a belief
    if brain.believes('distribution-beats-product'):
        ...

    # Get network
    experts = brain.network.domain_matches('sales')
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .loaders import (
    load_entities,
    load_relationships,
    load_predictions,
    load_agenda,
    load_network,
    load_state,
)
from .types import (
    ConfidenceLevel,
    RelationshipStrength,
    TrustLevel,
    Connection,
)


@dataclass
class NetworkAccessor:
    """Accessor for network queries."""

    _connections: list[dict] = field(default_factory=list)
    _stats: dict = field(default_factory=dict)

    @property
    def connections(self) -> list[dict]:
        return self._connections

    @property
    def stats(self) -> dict:
        return self._stats

    def domain_matches(self, domain: str) -> list[dict]:
        """Find connections in a domain."""
        domain_lower = domain.lower()
        return [
            c for c in self._connections
            if any(domain_lower in d.lower() for d in c.get('domains', []))
        ]

    def by_strength(self, strength: str) -> list[dict]:
        """Find connections by relationship strength."""
        return [
            c for c in self._connections
            if c.get('relationship_strength') == strength
        ]

    def high_trust(self) -> list[dict]:
        """Get high-trust connections."""
        return [c for c in self._connections if c.get('trust_level') == 'high']

    def energizing(self) -> list[dict]:
        """Get energizing connections."""
        return [c for c in self._connections if c.get('energy') == 'energizing']

    def draining(self) -> list[dict]:
        """Get draining connections."""
        return [c for c in self._connections if c.get('energy') == 'draining']

    def get(self, connection_id: str) -> Optional[dict]:
        """Get a specific connection."""
        for c in self._connections:
            if c.get('id') == connection_id:
                return c
        return None

    def search(self, query: str) -> list[dict]:
        """Search connections by name or company."""
        query_lower = query.lower()
        return [
            c for c in self._connections
            if query_lower in c.get('name', '').lower()
            or query_lower in (c.get('company') or '').lower()
        ]


class Brain:
    """
    Main Brain class for accessing shared intelligence.

    Usage:
        brain = Brain.load()
        if brain.believes('distribution-beats-product'):
            ...
    """

    def __init__(
        self,
        state: dict,
        entities: list,
        relationships: list,
        predictions: list,
        agenda: dict,
        network: dict,
    ):
        self._state = state
        self._entities = entities
        self._relationships = relationships
        self._predictions = predictions
        self._agenda = agenda
        self._network_data = network

    @classmethod
    def load(cls) -> Brain:
        """Load the brain from disk."""
        state = load_state()
        entities = load_entities()
        relationships = load_relationships()
        predictions = load_predictions()
        agenda = load_agenda()
        network = load_network()

        return cls(state, entities, relationships, predictions, agenda, network)

    # === STATE ===

    @property
    def state(self) -> dict:
        return self._state

    @property
    def version(self) -> str:
        return self._state.get('version', '')

    @property
    def capabilities(self) -> dict:
        return self._state.get('capabilities', {})

    def has_capability(self, name: str) -> bool:
        """Check if brain has a capability."""
        return self._state.get('capabilities', {}).get(name, False)

    # === ENTITIES ===

    @property
    def entities(self) -> list:
        return self._entities

    @property
    def beliefs(self) -> list:
        return [e for e in self._entities if e.get('type') == 'belief']

    @property
    def threads(self) -> list:
        return [e for e in self._entities if e.get('type') == 'thread']

    def entity(self, entity_id: str) -> Optional[dict]:
        """Get an entity by ID."""
        for e in self._entities:
            if e.get('id') == entity_id:
                return e
        return None

    def believes(self, belief_id: str) -> bool:
        """Check if the brain holds a belief."""
        full_id = f"belief.{belief_id}" if not belief_id.startswith('belief.') else belief_id
        return self.entity(full_id) is not None or self.entity(belief_id) is not None

    def belief_confidence(self, belief_id: str) -> Optional[str]:
        """Get belief confidence level."""
        full_id = f"belief.{belief_id}" if not belief_id.startswith('belief.') else belief_id
        belief = self.entity(full_id) or self.entity(belief_id)
        return belief.get('confidence') if belief else None

    def entities_by_type(self, entity_type: str) -> list:
        """Get entities by type."""
        return [e for e in self._entities if e.get('type') == entity_type]

    def entities_by_confidence(self, confidence: str) -> list:
        """Get entities by confidence level."""
        return [e for e in self._entities if e.get('confidence') == confidence]

    # === RELATIONSHIPS ===

    @property
    def relationships(self) -> list:
        return self._relationships

    def relationships_for(self, entity_id: str) -> list:
        """Get relationships for an entity."""
        return [
            r for r in self._relationships
            if r.get('from') == entity_id or r.get('to') == entity_id
        ]

    def supports(self, entity_id: str) -> list:
        """Get what supports an entity."""
        return [
            r for r in self._relationships
            if r.get('to') == entity_id and r.get('type') == 'supports'
        ]

    def contradicts(self, entity_id: str) -> list:
        """Get what contradicts an entity."""
        return [
            r for r in self._relationships
            if r.get('to') == entity_id and r.get('type') == 'contradicts'
        ]

    # === PREDICTIONS ===

    @property
    def predictions(self) -> list:
        return self._predictions

    def pending_predictions(self) -> list:
        """Get pending predictions."""
        return [p for p in self._predictions if p.get('status') == 'pending']

    def predictions_due_before(self, date_str: str) -> list:
        """Get predictions due before a date."""
        return [
            p for p in self._predictions
            if p.get('status') == 'pending' and p.get('resolution_date', '') <= date_str
        ]

    # === AGENDA ===

    @property
    def agenda(self) -> dict:
        return self._agenda

    def urgent_agenda_items(self) -> list:
        """Get high-priority agenda items."""
        return [
            item for item in self._agenda.get('immediate', [])
            if item.get('priority') == 'high'
        ]

    # === NETWORK ===

    @property
    def network(self) -> NetworkAccessor:
        return NetworkAccessor(
            _connections=self._network_data.get('connections', []),
            _stats=self._network_data.get('stats', {}),
        )

    # === CONVENIENCE ===

    @property
    def pending_attention(self) -> list:
        """Get attention items that need immediate focus."""
        return self._state.get('pending_attention', [])

    @property
    def recent_changes(self) -> list:
        """Get recent changes to the brain."""
        return self._state.get('recent_changes', [])
